### PR TITLE
Add recommendations regarding use of datapipes for multi-processing, shuffling, DDP, etc.

### PR DESCRIPTION
**Pull request by NicolasHug**
Closes https://github.com/pytorch/text/issues/1751

Rendered docs: https://output.circle-artifacts.com/output/job/62244176-8485-40e7-8a19-92d89b4a64a3/artifacts/0/docs/datasets.html (not fully up to date)

[Edit: Parmeet]: Rendered docs: https://output.circle-artifacts.com/output/job/a328cfc0-5824-4a1a-a3bf-630e09ff37ee/artifacts/0/docs/datasets.html

@parmeet @Nayef211 the rest of the week is a bank holiday for the UK, so I might not be able to address comments right away. Please feel free to directly edit this PR!

Also FYI @ejguan @NivekT @VitalyFedyunin

### CODE REVIEW

**Review by parmeet — COMMENTED** [review]
Thanks so much @NicolasHug for creating this PR. I think overall it LGTM. I just have few general questions in terms of Distributed training :).

File: docs/source/datasets.rst [hunk 1]
```diff
@@ -3,6 +3,51 @@ torchtext.datasets
...
+      are equal.
+
+    - Distributed training with ``DistributedDataParallel`` is not yet entirely
+      stable / supported, and we don't recommend it at this point. It will be
+      better supported in DataLoaderV2. If you still wish to use DDP, make sure
+      that:
+
+      - All workers (DDP workers *and* DataLoader workers) see a different part
+        of the data. You might need to call ``dp = dp.apply_sharding(world_size,
+        rank)``.
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply docs/source/datasets.rst:44]: @parmeet good catch - users would need to wrap the dp again before calling `apply_sharding()`. I'll update the comment

```py
from torchdata.datapipes.iter import IterableWrapper, ShardingFilter

a = IterableWrapper(range(30))

a = ShardingFilter(a)
a.apply_sharding(num_of_instances=2, instance_id=0)

a = ShardingFilter(a)  # without this, only the second sharding (5) applies
a.apply_sharding(num_of_instances=5, instance_id=0)

list(a)
# [0, 10, 20]
```

File: docs/source/datasets.rst [hunk 2]
```diff
@@ -3,6 +3,51 @@ torchtext.datasets
...
+      - The shuffling seed is the same across all workers. You might need to
+        call ``torch.utils.data.graph_settings.apply_shuffle_seed(dp, rng)``
+      - The suffling seed is different across epochs.
+      - The rest of the RNG (typically used for transformations) is
+        **different** across workers, for maximal entropy and optimal accuracy.
+
```
[the remark this answers is not in the document]
Comment (nateanl) [reply docs/source/datasets.rst:58]: It's not unusual to have different number of samples in each rank, for example, if the datapipe expects the same number of tokens/characters for each batch, each sentence may have different number of tokens/characters, then the batch size will be different. As long as the loss value is normalized in a correct way, the training should be fine.

[the remark this answers is not in the document]
Comment (nateanl) [reply docs/source/datasets.rst:58]: @NicolasHug Oh you are referring the number of batches in each rank. That's correct, DDP will hang forever if the number of batches are not even in all ranks.  I was talking about the batch_size and it can be different :)

[the remark this answers is not in the document]
Comment (NicolasHug) [reply docs/source/datasets.rst:57]: Thanks all, I added a section about all this just above

>  - All DDP workers work on the same number of batches. One way to do this is to by limit the size of the datapipe within each worker to ``len(datapipe) // num_ddp_workers``, but this might not suit all use-cases.

[the remark this answers is not in the document]
Comment (parmeet) [reply docs/source/datasets.rst:57]: yupp, SGTM!

File: docs/source/datasets.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (parmeet) [reply docs/source/datasets.rst]: I think this sounds neat. This would also avoid applying sharding filter multiple times. Given that datasets already wrapped the dp inside sharding filter, all we need to do is provide the implementation of worker_init_fn that does sharding according to world size and worker IDs. 

```python
def worker_init_fn(worker_id, world_size, rank):
    info = torch.utils.data.get_worker_info()
    num_workers = info.num_workers
    datapipe = info.dataset
    datapipe.apply_sharding(world_size * worker_num, rank * worker_num + worker_id)

# Initialize DataLoader within each rank with above function
dl = DataLoader(dataset, worker_init_fn=functools.partial(worker_init_fn, world_size=world_size, rank=rank), ...)
```

Does this sounds about right @ejguan?

[the remark this answers is not in the document]
Comment (NicolasHug) [reply docs/source/datasets.rst]: Probably best to get work_size and rank from within the function? One can use `int(os.environ["RANK"])` and `int(os.environ["WORLD_SIZE"])`. Also for clarity maybe we could rename these to `num_ddp_workers`, `num_dataloader_workers`, `ddp_worker_id`, `dataloader_worker_id`.

It might be worth documenting both ways? The second one is somehow more complex, because it exposes the DataLoader sharding level.

[the remark this answers is not in the document]
Comment (ejguan) [reply docs/source/datasets.rst]: @parmeet That is the solution in my mind.

@NicolasHug That would be the ideal case. I believe worker processes inherit the environment variable like `RANK` and `WORLD_SIZE`. But, considering that I only tested it in multiprocessing environment and my bare knowledge on distributed processes, I was hesitating to recommend to fetch such env in worker processes.

[the remark this answers is not in the document]
Comment (parmeet) [reply docs/source/datasets.rst]: SGTM! I think we can leave the details of implementation to user. At higher level, probably it is sufficient to communicate the workaround with worker_init_fn :)

[the remark this answers is not in the document]
Comment (VitalyFedyunin) [reply docs/source/datasets.rst]: Please do not use ENV variables. There are built-in things for it:
```python
import torch.distributed as dist
if dist.is_available():
        total_workers *= dist.get_world_size()
        global_worker_id = dist.get_rank()*info.num_workers + global_worker_id
```

I will submit PR(s) to apply it automatically this week.

The only remaining problem is the possibility of having a different number of processes on different ranks. I prefer to keep one `.sharding_filter` for it, but perhaps would have to introduce sharing ranks/levels into it.

[the remark this answers is not in the document]
Comment (parmeet) [reply docs/source/datasets.rst]: Thanks @VitalyFedyunin for the feedback. Just so I understand, do you mean to submit PR for updating [worker_init_fn](https://github.com/pytorch/pytorch/blob/d90012689f5b99bbc289da598ad982f829687288/torch/utils/data/backward_compatibility.py#L4) such that it not only works for multi-processing case (as of now) but shall also take into account distributed settings like the one propose here https://github.com/pytorch/text/pull/1755#discussion_r886909628?

[the remark this answers is not in the document]
Comment (VitalyFedyunin) [reply docs/source/datasets.rst]: Yes.  And also classical DataLoader will automatically add this worker_init_fn pieces to all DataPipe-based Datasets. https://github.com/pytorch/pytorch/pull/78631

File: docs/source/datasets.rst [hunk 3]
```diff
@@ -3,6 +3,58 @@ torchtext.datasets
...
+      stable / supported, and we don't recommend it at this point. It will be
+      better supported in DataLoaderV2. If you still wish to use DDP, make sure
+      that:
+
+      - All workers (DDP workers *and* DataLoader workers) see a different part
+        of the data. You might need to call ``dp = dp.apply_sharding(world_size,
+        rank)`` after wrapping the datapipe into a `ShardingFilter
+        <https://pytorch.org/data/main/generated/torchdata.datapipes.iter.ShardingFilter.html>`_.
+        You'll want to do that early in the datapipe, to avoid needlessly
+        processing samples that eventually get dropped by the workers.
```
Comment (parmeet) [inline docs/source/datasets.rst:47]: @NicolasHug I would suggest we wait for @VitalyFedyunin to land PRs that would take care of sharding automatically through usage of worker_init_fn and we can remove this particular point. As per the issue https://github.com/pytorch/text/issues/1727, we have already wrapped datasets in sharding filter, which means that the users do not have to apply it explicitly nor have to take into account DataLoaderV1+worker_init_fn mechanics :).

File: docs/source/datasets.rst [hunk 4]
```diff
@@ -3,6 +3,61 @@ torchtext.datasets
...
+
+      - All workers (DDP workers *and* DataLoader workers) see a different part
+        of the data. The datasets are already wrapped inside  `ShardingFilter
+        <https://pytorch.org/data/main/generated/torchdata.datapipes.iter.ShardingFilter.html>`_
+        and you may need to call ``dp.apply_sharing(num_shards, shard_id)`` in order to shard the 
+        data across ranks (DDP workers) and DataLoader workers. One way to do this
+        is to create ``worker_init_fn`` that calls ``apply_sharding`` with appropriate 
+        number of shards (DDP workers * DataLoader workers) and shard id (inferred through rank
+        and worker ID of corresponding DataLoader withing rank). Note however, that this assumes
+        equal number of DataLoader workers for all the ranks.
```
Comment (parmeet) [inline docs/source/datasets.rst:50]: @NicolasHug, @VitalyFedyunin , @ejguan : I made the description rather high level while leaving implementation details to the users. My thinking here is to make the user aware of pitfalls without necessarily providing specific recommendation on implementing specific solution.
